src/crystal/system/file.cr

```diff
@@ -1,8 +1,34 @@
+require "file/mode"
+
 # :nodoc:
 module Crystal::System::File
+  private ModeMap = {
```

Comment:
I'd prefer this not to be implemented with a hash, it's slow. Manual logic is better here:

```cr
flags |= O_CREAT if mode.create?
```

Reply:
Aaaaaasssssss yoooouuuuu wwissssshhhhhh.

Reply:
I just noticed you had two comments.

> I'd rather have `:read, :write` than `:readwrite`. That's what I thought we agreed on in #7857 

Ok.

> Using `:append` with `:write` should be an error.

Not harmless?  Ok.

> And using `:sync` should probably set `IO#sync`??? Not sure on the last one.


Applications using sync would probably flush at specific points and possibly want `IO` buffering to reduce write load no?

```
io.write_bytes frame_length
io.write data
io.write_bytes checksum
io.flush
```

> And it needs a `*mode : Mode` overload which allows calling `File.open("foo", :read, :write)`.

How does that work with additional args?  Which of these if any should I use?
```
def self.new(filename : Path | String, mode : Mode = File::Mode.flags(Read), perm = DEFAULT_CREATE_PERMISSIONS, encoding = nil, invalid = nil)
def self.new(filename : Path | String, mode : Mode = File::Mode.flags(Read), *, perm = DEFAULT_CREATE_PERMISSIONS, encoding = nil, invalid = nil)
def self.new(filename : Path | String, mode : Mode = File::Mode.flags(Read))
```

Reply:
You want to add a new

```cr
def self.new(filename : Path | String, *modes : Mode, perm = DEFAULT_CREATE_PERMISSIONS, encoding = nil, invalid = nil)
  new(filename, modes.reduce(&.|), encoding, invalid)
end
```

Reply:
> Applications using sync would probably flush at specific points and possibly want `IO` buffering to reduce write load no?

yeah that's what I thought, let's not do this

Reply:
> Not harmless? Ok.

It's just a bit confusing behaviour, is all. Since writes will all be at the end but reads won't?

I'm not 100% sure about disallowing it, it just seems confusing behaviour for the end user if they accidentally do this. And if it's got a usecase, then we can easily allow it in the future.

Reply:
You said "Using :append with :write should be an error.".  Now your talking about read. Did you mean :read with :append?

Reply:
Splats are new to me.  I have two errors I'm not sure how to resolve.

`File.new "foo", :read, :write`

```
In src/file.cr:111:62

 111 | fd = Crystal::System::File.open(filename, modes.reduce(&.|), perm)
                                                                ^
Error: wrong number of arguments for 'File::Mode#|' (given 0, expected 1)
```


`File.new "foo", :read, :write, :create_new` 

```
In src/file.cr:78:10

 78 | self.set_encoding(encoding, invalid: invalid) if encoding
           ^-----------
Error: no overload matches 'File#set_encoding' with types Symbol, invalid: Nil
```

I tried `[:create_new, :createnew, :CreateNew, :Create_New]`.  Do I need to type restrict `encoding` and `invalid`?  Are enum splats broken with camelcase?

Reply:
> Did you mean :read with :append?

Shit, my bad, yes I did.

Reply:
A replicated transaction log may want :read with :append.  One writer appends.  Clients dynamically connect and start streaming from an arbitrary point.  Kafka uses this design.  Many database replication implementations probably do something similar.

Reply:
Okay, we can allow it.

BTW, `O_RDONLY | O_WRONLY != O_RDWR`, which this code currently assumes.

I'll work on a workaround for the bug above.

Reply:
A replicated transaction log may use :read with :append.  One writer appends.  Clients dynamically connect and streaming from an arbitrary point.  Kafka uses this design.  Many database replication implementations probably do something similar.

Reply:
I assumed you were busy in the past week so I did my own.